cloudimg/ms_azure.py

```diff
@@ -0,0 +1,513 @@
+            log.info(str(self._upload_progress))
+
+        return self._upload_progress
+
+    def upload_to_container(self, image_path, container_name, object_name,
+                            tags, **kwargs):
+        """
+        Upload a blob to a container in a storage account and tag it.
+        If the file_url starts with http, use the upload_blob_from_url method,
+        otherwise use the upload_blob method.
+
+        Args:
+            image_path (str): Local or remote path to the image
+            container_name (str): The container to upload to
+            object_name (str): The uploaded file name
+            tags (dict): Tags to apply to the uploaded image
+            **kwargs: Additional arguments for upload_blob and
+                      upload_blob_from_url
+
+        Returns:
+            The uploaded BlobClient on success.
+        Raises:
+            AzureError on upload failure.
+        """
+        # Initial parameters
+        self._upload_progress = UploadProgress(current=0, total=0)
+        self._upload_last_log = self._upload_progress.last_seen
+        upload_default_args = {
+            "tags": tags,
+            "max_concurrency": self.UPLOAD_MAX_CONCURRENCY,
+            "raw_response_hook": self.upload_callback,
+        }
+        for k, v in upload_default_args.items():
+            kwargs.setdefault(k, v)
+
+        # Setup the destination container
+        container_client = self.get_container_by_name(
+            name=container_name,
+            create=True
+        )
+
+        # Check the image tags
+        log.info("Checking whether the image was already uploaded.")
+        if self.are_tags_present(container_client, tags):
+            log.info(
+                "The tags '%s' already exists in storage account "
+                "'%s' under container "
+                "'%s' \n"
+                "thus the image will not be uploaded." %
+                (
+                    tags,
+                    container_client.account_name,
+                    container_client.container_name
+                )
+            )
+            return
+
+        # Upload to container and tag image
+        blob_client = container_client.get_blob_client(blob=object_name)
+        if blob_client.exists():
+            log.info(
+                "The image \"%s\" already exists in container \"%s\"",
+                object_name,
+                container_name,
+            )
+            return
+
+        log.info('Uploading %s to container %s', image_path, container_name)
+        log.info('Uploading %s with name %s', image_path, object_name)
+        if image_path.startswith("http:") or image_path.startswith("https:"):
+            blob_client.upload_blob_from_url(source_url=image_path, **kwargs)
+        else:
+            with open(image_path, "rb") as data:
+                blob_client.upload_blob(data=data, **kwargs)
+        log.info(self._upload_progress)
```

Comment:
There's a bit of inconsistency here, as elsewhere this has been written:

```
log.info(str(self._upload_progress))
```

So some of the code explicitly converts the upload progress object to a string, while other code assumes that the info method is going to do it for you.  They'll both work but it'd be nice to be consistent.

Reply:
I've changed it to be consistent